src/peft/tuners/vera/model.py

```diff
@@ -0,0 +1,549 @@
+        ...     "kakaobrain/kogpt",
+        ...     revision="KoGPT6B-ryan1.5b-float16",  # or float32 version: revision=KoGPT6B-ryan1.5b
+        ...     pad_token_id=tokenizer.eos_token_id,
+        ...     use_cache=False,
+        ...     device_map={"": rank},
+        ...     torch_dtype=torch.float16,
+        ...     load_in_8bit=True,
+        ... )
+        >>> model = prepare_model_for_int8_training(model)
+        >>> vera_model = get_peft_model(model, config)
+        ```
+
+    **Attributes**:
+        - **model** ([`~transformers.PreTrainedModel`]) -- The model to be adapted.
+        - **peft_config** ([`VeraConfig`]): The configuration of the Vera model.
+    """
+
+    def _find_first_dim(self) -> int:
+        """
+        Finds the first linear and embedding that has been wrapped with Vera, and extract the input and output
+        dimension.
+
+        This will be used for determining the size of the shared vera_A and vera_B matrices.
+
+        This will throw an error if there are multiple layers of the same type with different shapes.
+        """
+        first_linear, first_embedding = None, None
+        for module in self.model.modules():
+            if isinstance(module, Linear):
+                module_shape = tuple(module.weight.shape)
+                if module.fan_in_fan_out:
+                    module_shape = module_shape[::-1]
+
+                if first_linear is not None and module_shape != first_linear:
+                    raise ValueError(
+                        f"Multiple target linear layers with different dimensions were specified! Vera only supports a single dimension size. Got '{module_shape}' expected '{first_linear}"
+                    )
+                first_linear = module_shape
+
+            elif isinstance(module, Embedding):
+                if first_embedding is not None and tuple(module.weight.shape) != first_embedding:
+                    raise ValueError(
+                        "Multiple target embedding layers with different dimensions or vocabulary sizes were specified! Vera only supports a single size."
+                    )
+                first_embedding = tuple(module.weight.shape)
+
+            if first_linear is not None and first_embedding is not None:
+                return first_linear, first_embedding
+
+        return first_linear, first_embedding
+
+    def __init__(self, model, config, adapter_name) -> None:
+        super().__init__(model, config, adapter_name)
+        config = config[adapter_name]
+
+        first_linear, first_embedding = self._find_first_dim()
+
+        if first_embedding is not None:
+            first_embedding_vocab_size, first_embedding_dim = first_embedding
+        if first_linear is not None:
+            first_linear_out_dim, first_linear_in_dim = first_linear
+
+        # deterministic init of vera_A and vera_B if we know the key
+        generator = torch.Generator(device="cpu").manual_seed(config.projection_prng_key)
```

Comment:
What if `projection_prng_key` is None?

Reply:
this should have been caught in the config init, but I can add an additional check here if you want

Reply:
Where do we check that? The type is `Optional`.

Reply:
It is `Optional` and defaults to `None` as I couldn't really think of a sensible default for a PRNG key. However, actually passing `None` should result in an error.

I thought it was checked in `config.py` but I seem to be have been mistaken. Therefore, I will add checks for it in `model.py`.